### PR TITLE
A service message will not be sent from iframe if it is removed afterwards (fix #64)

### LINKED ISSUE #64: A service message will not be sent from iframe if it is removed afterwards
A service message will not be sent from iframe if it is removed afterwards
Our on-farm functional tests in the default branch fail in Chrome due to this error.

1) in the latest version of Chrome, the timeout is reset (even the window.top timeout that was set after the iframe was unloaded)
sendServiceMsg method:
```js
 NativeMethods.setTimeout.call(topWindow, function () {
            targetWindow[RECEIVE_MSG_FN]({
                data:   JSON.parse(JSON.stringify(message)), 
                source: window
            });
        }, 10);
```
2)  in IE, a parse error is raised after the iframe is removed
```js
JSON.parse(JSON.stringify(message))
```
Regression qunit test:
```js
asyncTest('service message from removed iframe', function () {
    var $iFrame         = $('<iframe>');
    var messageReceived = false;

    $iFrame[0].src = window.QUnitGlobals.getResourceUrl('../../data/same-domain/service-message-from-removed-iframe.html');
    $iFrame.appendTo('body');

    window.setTimeout(function(){
        ok(messageReceived, 'message is not received');
        start();
    }, 100);

    MessageSandbox.on(MessageSandbox.SERVICE_MSG_RECEIVED, function () {
        messageReceived = true;
    });
});
```
tested iframe markup:
```js
<!DOCTYPE html>
<html>
<head>
    <title></title>
    <meta charset="utf-8">
    <script src="/hammerhead.js" class="script-TC2b9a6d"></script>
</head>
<body>
<script type="text/javascript">
    var hammerhead = window.Hammerhead;
    var urlUtil = hammerhead.get('./utils/url');

    hammerhead.get('./settings').set({
        CROSS_DOMAIN_PROXY_PORT: 2000
    });

    urlUtil.OriginLocation.get = function () {
        return 'https://example.com';
    };

    var MessageSandbox = Hammerhead.get('./sandboxes/message');
    var scriptProcessor = Hammerhead.get('../processing/script');

    Hammerhead.init();

    var iframe = window.top.document.getElementsByTagName('iframe')[0];

    debugger;
    MessageSandbox.sendServiceMsg('', window.top);

    iframe.parentNode.removeChild(iframe);
</script>
<div></div>
</body>
</html>
```

### PR DESCRIPTION
**Pull request by LavrovArtem**
/cc @churkin @miherlosev

### CODE REVIEW

File: src/client/sandbox/event/message.js [hunk 1]
```diff
@@ -33,9 +33,12 @@ export default class MessageSandbox extends SandboxBase {
         this.topWindow = null;
         this.window    = null;
 
-        this.listeners = listeners;
+        this.listeners     = listeners;
+        this.unloadSandbox = unloadSandbox;
 
         this.storedOnMessageHandler = null;
+
+        this.serviceMsgsQueue = [];
```
Comment (inikulin) [inline src/client/sandbox/event/message.js]: Is this client-server service messages or cross-domain communication service messages?
If first, we should move this code to `transport`, if second we should rename them somehow to distinct them from client-server XHR methods.

**Comment by inikulin** [comment]
\r-

File: src/client/sandbox/event/message.js (hunk 1 above)
Comment (LavrovArtem) [inline src/client/sandbox/event/message.js]: Is this service messages between iframe without src and same-domain iframe

Comment (inikulin) [inline src/client/sandbox/event/message.js]: Then we need to rename it. Discuss naming with @VasilyStrelyaev

Comment (LavrovArtem) [inline src/client/sandbox/event/message.js]: maybe ```serviceMsgsBetweenSameDomainIframe```

Comment (VasilyStrelyaev) [inline src/client/sandbox/event/message.js]: Won't `interIframeMsgQueue` be enough?

Comment (LavrovArtem) [inline src/client/sandbox/event/message.js]: this name does not show that applies only to same-domain without cross-domain

Comment (VasilyStrelyaev) [inline src/client/sandbox/event/message.js]: should it? is there something that can be confused with it?

Comment (VasilyStrelyaev) [inline src/client/sandbox/event/message.js]: We shouldn't use `interIframeSameDomainMsgQueue` unless it's super-necessary.

**Comment by LavrovArtem** [comment]
FPR

File: src/client/sandbox/event/message.js [hunk 2]
```diff
@@ -33,9 +33,12 @@ export default class MessageSandbox extends SandboxBase {
         this.topWindow = null;
         this.window    = null;
 
-        this.listeners = listeners;
+        this.listeners     = listeners;
+        this.unloadSandbox = unloadSandbox;
 
         this.storedOnMessageHandler = null;
+
+        this.interIframeMsgQueue = [];
```
Comment (inikulin) [inline src/client/sandbox/event/message.js]: It's better be named `iframeInternalMsgQueue

File: src/client/sandbox/event/message.js [hunk 3]
```diff
@@ -105,11 +108,32 @@ export default class MessageSandbox extends SandboxBase {
         return this.window.top === this.window.self && this.window !== this.topWindow;
     }
 
+    _removeServiceMsgFromQueue (sendFunc) {
```
Comment (inikulin) [inline src/client/sandbox/event/message.js]: `_removeInternalMsgFromQueue`

**Comment by inikulin** [comment]
\r-

**Comment by LavrovArtem** [comment]
FPR

**Comment by inikulin** [comment]
lgtm

**Comment by miherlosev** [comment]
lgtm

**Comment by churkin** [comment]
lgtm

**Pull request merged**